crates/router/src/core/payments/operations/payment_confirm.rs

```diff
@@ -718,6 +756,174 @@ impl<F: Clone, Ctx: PaymentMethodRetrieve>
+                .and_then(|pm_data| match pm_data {
+                    api_models::payments::PaymentMethodData::Card(card) => {
+                        crypto::HmacSha512::sign_message(
+                            &crypto::HmacSha512,
+                            merchant_secret.as_bytes(),
+                            card.card_number.clone().get_card_isin().as_bytes(),
+                        )
+                        .change_context(errors::ApiErrorResponse::InternalServerError)
+                        .attach_printable("error in card bin hash creation")
+                        .map_or_else(
+                            |err| {
+                                logger::error!(error=?err);
+                                None
+                            },
+                            Some,
+                        )
+                    }
+                    _ => None,
+                });
+
+            // Hashed Extended Cardbin to check whether or not this payment should be blocked.
+            let extended_cardbin_hash =
+                payment_data
+                    .payment_method_data
+                    .as_ref()
+                    .and_then(|pm_data| match pm_data {
+                        api_models::payments::PaymentMethodData::Card(card) => {
+                            crypto::HmacSha512::sign_message(
+                                &crypto::HmacSha512,
+                                merchant_secret.as_bytes(),
+                                card.card_number.clone().get_extended_card_bin().as_bytes(),
+                            )
+                            .change_context(errors::ApiErrorResponse::InternalServerError)
+                            .attach_printable("error in extended card bin hash creation")
+                            .map_or_else(
+                                |err| {
+                                    logger::error!(error=?err);
+                                    None
+                                },
+                                Some,
+                            )
+                        }
+                        _ => None,
+                    });
+
+            let cardbin_encoded_hash = cardbin_hash.map(|id| consts::BASE64_ENGINE.encode(id));
+
+            let extended_cardbin_encoded_hash =
+                extended_cardbin_hash.map(|id| consts::BASE64_ENGINE.encode(id));
+
+            let fingerprint_encoded_hash =
+                fingerprint_hash.map(|id| consts::BASE64_ENGINE.encode(id));
+            let mut fingerprint_id = None;
+
+            //validating the payment method.
+            let mut is_pm_blocklisted = false;
+            let find_for_fingerprint = db
+                .find_blocklist_lookup_entry_by_merchant_id_kms_decrypted_hash(
+                    merchant_id.clone(),
+                    fingerprint_encoded_hash
+                        .clone()
+                        .ok_or(errors::ApiErrorResponse::InternalServerError)?,
```

Comment:
this `.ok_or` will always error for payment methods other than card. Instead gate this logic with an `if let Some()`. Same for the other two.